src/base/bittorrent/ltqbitarray.cpp

```diff
@@ -0,0 +1,73 @@
+/*
+ * Bittorrent Client using Qt and libtorrent.
+ * Copyright (C) 2022  Mike Tzou (Chocobo1)
+ *
+ * This program is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU General Public License
+ * as published by the Free Software Foundation; either version 2
+ * of the License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ *
+ * In addition, as a special exception, the copyright holders give permission to
+ * link this program with the OpenSSL project's "OpenSSL" library (or with
+ * modified versions of it that use the same license as the "OpenSSL" library),
+ * and distribute the linked executables. You must obey the GNU General Public
+ * License in all respects for all of the code used other than "OpenSSL".  If you
+ * modify file(s), you may extend this exception to your version of the file(s),
+ * but you are not obligated to do so. If you do not wish to do so, delete this
+ * exception statement from your version.
+ */
+
+#include "ltqbitarray.h"
+
+#include <memory>
+
+#include <libtorrent/bitfield.hpp>
+
+#include <QBitArray>
+
+namespace
+{
+    unsigned char reverseByte(const unsigned char byte)
+    {
+        // https://graphics.stanford.edu/~seander/bithacks.html#ReverseByteWith64Bits
+        return (((byte * 0x80200802ULL) & 0x0884422110ULL) * 0x0101010101ULL) >> 32;
+    }
+}
+
+namespace BitTorrent::LT
+{
+    QBitArray toQBitArray(const lt::bitfield &bits)
+    {
+        const int STACK_ALLOC_SIZE = 10 * 1024;
+
+        const char *bitsData = bits.data();
+        const int dataLength = (bits.size() + 7) / 8;
+
+        if (dataLength <= STACK_ALLOC_SIZE)
+        {
+            // fast path for small bitfields
+            char tmp[STACK_ALLOC_SIZE];  // uninitialized for faster allocation
```

Comment:
Maybe split extract two implementations and call appropriate one? Otherwise stack will be allocated independently from bitfield size.

Reply:
would wrapping the fast path into a lambda suffice?

Reply:
> would wrapping the fast path into a lambda suffice?

I think Yes. Data will be allocated on stack only when lambda is called.

Reply:
Unfortunately wrapping in a lambda isn't without overhead, the overhead just cancels out the gains of allocating on stack and now the fast path is as slow as the slow path (for 10000 pieces).
Maybe just keep it as is? The overhead of additional memory on stack is just negligible (in both space and time wise).

Reply:
>Maybe just keep it as is? The overhead of additional memory on stack is just negligible (in both stack space and run time wise).

But why don't try use regular function instead of lambda?


> now the fast path is as slow as the slow path (for 10000 pieces).

IIRC, the number of pieces should not affect, since they differ only in the way memory is allocated, and then the code is executed the same way.
BTW, I wonder how significant is the difference between them? And how many cases (torrents) are covered by chosen STACK_ALLOC_SIZE. If it goes this way (fast code path) in the vast majority of cases, then you can leave it as it is.

Reply:
>But why don't try use regular function instead of lambda?

Regular function is faster than lambda but still not the best we can get here. I don't really know what your concern is about.

>BTW, I wonder how significant is the difference between them? 

The numbers fluctuates it is hard to be precise here. On average, the current impl is around  7200 ns, lambda version is around 8100 ns and function version is around 7300 ns. This is measuring `toQBitArray()` as a whole.

>And how many cases (torrents) are covered by chosen STACK_ALLOC_SIZE. 

You can't really answer it can you?
The size is chosen to be as large as possible and it still leaves a safe margin that won't overflow the stack, that's all.